### DISCUSSION TITLE
where's Linux?

### DISCUSSION THREAD

**Discussion by sanana-banana** [Anjok07/ultimatevocalremovergui#579] [General]
I can't find the Linux download, is there something I'm missing?

**Comment by ste1590** [comment]
https://github.com/Anjok07/ultimatevocalremovergui/archive/refs/heads/master.zip

**Reply by sanana-banana** [reply]
oh why thank you

**Reply by sanana-banana** [reply]
um excuse me but how do I install one of these

**Comment by ste1590** [comment]
Via pip3, you can find instructions for Ubuntu in the readme file. If you use another distribution, then use the appropriate package manager for that distribution before the last command "pip3 install -r requirements.txt". Currently, Ubuntu 22.04 / 22.10 works well, the same for Linux Mint 21 and probably Fedora 37. If you are using a rolling release distro, Arch or Debian Sid based this version will not work for a while due to some Python dependencies. Also, if you manage to install this and get a choppy sound in your audio file after all, just remove gstreamer1.0-libav package from the system. It's currently bugged in Ubuntu.

**Reply by sanana-banana** [reply]
oh thank you

**Comment by ste1590** [comment]
Just an update on this regarding the Linux install. In my previous answer, I wrote that due to Python dependencies, it will not work on rolling release distributions. Well, it will work like a charm on every distribution and version, I just tested this on Debian Sid and Arch as well. All you have to do is to install pyenv package and create a virtual Python environment for this purpose. The last version of Python that works with UVR 5.5.1 is 3.10.12. These are the installation steps for Debian, for other distributions just use the appropriate package manager but everything else is the same.

`sudo apt install ffmpeg`
`sudo apt install python3-pip`
`sudo apt install python3-tk`
`sudo apt install build-essential`
`sudo apt install libssl-dev libsqlite3-dev libbz2-dev libgdbm-dev libncurses5-dev libncursesw5-dev libreadline-dev zlib1g-dev libffi-dev tk-dev lzma liblzma-dev`
`sudo apt install curl`
`sudo apt install git`
`curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash`
`echo 'export PATH="$HOME/.pyenv/bin:$PATH"' >> ~/.bashrc`
`echo 'eval "$(pyenv init -)"' >> ~/.bashrc`
`echo 'eval "$(pyenv virtualenv-init -)"' >> ~/.bashrc`
`source ~/.bashrc`
`pyenv install 3.10.12`
`pyenv global 3.10.12`
`pyenv local 3.10.12`

Now you need to download UVR from this link
`https://github.com/Anjok07/ultimatevocalremovergui/archive/refs/heads/master.zip`

The archive must be unzipped, and you need to cd to the directory and run the command:
`pip3 install -r requirements.txt`

At the very end, you can create a shortcut or just run the application from the terminal using this command (depending on where you unzip the UVR folder, I unzipped the archive in /opt, and I can run it like this from the terminal)

`/home/USERNAME/.pyenv/versions/3.10.12/bin/python3 /opt/ultimatevocalremovergui-master/UVR.py`

In case you need the system version of Python for something else, you can set it from the terminal with
`pyenv global system`
`pyenv local system`

UVR will still work as we have version 3.10.12 in the shortcut path and will not interfere with the system version of Python.

**Reply by harm6** [reply]
I tried to follow your advice, alas
`pyenv install 3.10.12`
returns:
```
...
Installing Python-3.10.12...
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/harm/.pyenv/versions/3.10.12/lib/python3.10/tkinter/__init__.py", line 37, in <module>
    import _tkinter # If this fails your Python may not be configured for Tk
ModuleNotFoundError: No module named '_tkinter'
WARNING: The Python tkinter extension was not compiled and GUI subsystem has been detected. Missing the Tk toolkit?
Installed Python-3.10.12 to /home/harm/.pyenv/versions/3.10.12
...
```
Any hint how to proceed?

Anyway, I did the other steps as well.  
Though,
`pip3 install -r requirements.txt`
caused a complete system crash while doing the torch-install...

:(

**Comment by ste1590** [comment]
What distribution are you using? Before you start with Python virtual env, make sure you installed all packages I mentioned above. The error you are reporting is related to the python3-tk package. If pip3 install command caused a complete system crash, it means that you are not using a virtual env instead of the system version of Python. Before you execute the command pip3 install, you need to set

`pyenv global 3.10.12`
`pyenv local 3.10.12`

**Reply by harm6** [reply]
Thanks for your reply!
> What distribution are you using?

Ubuntu 22.04.2
> The error you are reporting is related to the python3-tk package.

Yep, though, I don't know how to fix it in the virtual env....

> If pip3 install command caused a complete system crash, it means that you are not using a virtual env

Well
`$ which pip3`
returns
`/home/harm/.pyenv/shims/pip3`
Similar result for python, python3, pip 
For me it looks I _am_ using python from the virtual env.
What am I missing?

**Reply by ste1590** [reply]
Looks like Ubuntu needs tk-dev lzma and liblzma-dev packages.

`sudo apt install tk-dev lzma liblzma-dev`

Then
`pyenv global system`
`pyenv local system`
`pyenv uninstall 3.10.12`
`pyenv install 3.10.12`
`pyenv global 3.10.12`
`pyenv local 3.10.12`
cd to UVR dir and then  `pip3 install -r requirements.txt`

If you get an error regarding sklearn-pypi or scikit-learn during `pip3 install -r requirements.txt` you will need to remove folders sklearn and scikit_learn from /home/.local/lib/python3.10/site-packages, then you need to run the command again.

It should work. Please try and let me know.

**Reply by harm6** [reply]
I followed your advice step by step, alas `pip3 install -r requirements.txt` caused a system crash again.
Though I think I found the culprit: From an earlier attempt I had torch-2.0.1 installed. Obviously trying to install torch-1.13.1 without removing the torch-2.0.1 does no good.  Well, handling this more gracefully would have been nice...

Anyway, I have uvr up and running. I may have other usage-problems...
Is there a forum or a mailinglist apart from posting here at github?

Many thanks !!

**Reply by ste1590** [reply]
I am not sure what exactly caused the system crash in your case, I tried to recreate a similar issue, but I didn't succeed. Also, I tried on freshly installed Ubuntu 22.04.2 via VirtualBox and everything works fine. I personally never liked Ubuntu, Debian was always my first choice but anyway, it should work on any distro. Regarding the forum, I don't really know because I am not involved in this project or with the developers. I'm just an end user and a person who likes troubleshooting, so I just wanted to share some useful information and make it easier for other Linux users. :)

![uvr](https://github.com/Anjok07/ultimatevocalremovergui/assets/95093618/dd4c979b-96d1-4a08-88e1-329c14867c78)

**Comment by joshgura** [comment]
arch, using a conda env:

important things i learned while trying to install on linux so far:

remove dora from requirements.txt otherwise installation of requirements is not possible

<s>when creating a conda env, use python 3.11 otherwise get a pyglet error.</s>

reinstalling with these conditions. will report if it worked.

no, it didn't work. got the pyglet playsound error.

i see above that people are using python 3.10.12?  is that the magic bullet? trying that. will report what happens.

edit: no, after wiping my old env, making a new one for python 3.10.12, still getting the "playsound is relying on another python subprocess" i wonder what that subprocess is supposed to be. pyglet etc etc.

**Comment by ste1590** [comment]
Follow the steps carefully I wrote above, it still works fine on Debian. Try using pyenv as well.

**Reply by joshgura** [reply]
I appreciate it, but many of those packages don't exist on arch. there's another guy around here on arch who claimed to get uvr to run so i haven't given up hope.  I'm using conda to handle virtual environments, there's no difference in how they perform.

also, on arch these packages don't exist under these labels, (but I already have them installed because i probably install and try all the killer apps)  libssl-dev libsqlite3-dev libbz2-dev libgdbm-dev libncurses5-dev libncursesw5-dev libreadline-dev zlib1g-dev 

python3-tk doesn't exist in the package manager or the aur. (or pip for that matter, it's just tk)

the 'playsound and pyglet' error isn't uncommon. but really UVR is only theoretically working on linux, for those who do have it working, it's hanging on by the skin of the teeth.

**Reply by ste1590** [reply]
Indeed, something is wrong. I installed Arch on VirtualBox, but I can't install UVR anymore either. Let me figure this out, and I'll let you know if I find a solution.

**Comment by ste1590** [comment]
Updated installation guide for Arch and current UVR, master branch.

sudo pacman -Syu ffmpeg
sudo pacman -Syu python-pip
sudo pacman -Syu tk
sudo pacman -Syu base-devel
sudo pacman -Syu glu
sudo pacman -Syu curl
sudo pacman -Syu git
curl -L https://github.com/pyenv/pyenv-installer/raw/master/bin/pyenv-installer | bash
echo 'export PATH="$HOME/.pyenv/bin:$PATH"' >> ~/.bashrc
echo 'eval "$(pyenv init -)"' >> ~/.bashrc
echo 'eval "$(pyenv virtualenv-init -)"' >> ~/.bashrc
source ~/.bashrc
pyenv install 3.11.10
pyenv global 3.11.10
pyenv local 3.11.10
pip install --upgrade pip
pip install --upgrade setuptools wheel
pip install pygobject

Now you need to download UVR from this link
https://github.com/Anjok07/ultimatevocalremovergui/archive/refs/heads/master.zip

The archive must be unzipped, and you need to cd to the directory and run the command:

Before pip3 install, open requirements.txt file and remove line Dora==0.0.3
Now we can run from terminal:
pip3 install -r requirements.txt

At the very end, you can create a shortcut or just run the application from the terminal using this command (depending on where you unzip the UVR folder, I unzipped the archive in /opt, and I can run it like this from the terminal)

/home/USERNAME/.pyenv/versions/3.11.10/bin/python3 /opt/ultimatevocalremovergui-master/UVR.py

In case you need the system version of Python for something else, you can set it from the terminal with
pyenv global system
pyenv local system

![uvr](https://github.com/user-attachments/assets/f1859427-f617-4013-95a5-a0cb2ccc4a7a)

**Reply by joshgura** [reply]
Thanks! I think -Syu upgrades all packages globally, not just the ones in question. -Syu should only really be needed once since all the following times are redundant unless a new package somewhere in the global system got released every few seconds.

still getting the playsound is relying on another python subprocess, please use 'pip install pygobject' but:

`Requirement already satisfied: pygobject in ./opt/UVR/env/lib/python3.10/site-packages (3.50.0)
Requirement already satisfied: pycairo>=1.16 in ./opt/UVR/env/lib/python3.10/site-packages (from pygobject) (1.27.0)
`

the last error in the pyglet errors is:

`  File "/home/username/opt/UVR/env/lib/python3.10/site-packages/pyglet/gl/glx_info.py", line 86, in have_version
    client_version = self.get_client_version().split()[0]
IndexError: list index out of range
`

a lot of graphics related words in the pyglet part, so i thought it was tk, but tk is installed.
also, i installed under python 3.11 but there's no change to the error message.

i'm inquiring at the pyglet developer and at the arch forum. could be that opengl is running on my cpu but pyglet is expecting it to be with the gpu.

I FIGURED SOMETHING OUT

opengl is currently being rendered by my integrated graphics and not my nvidia gpu, so if I put the following before the python command I get a GUI now! (it's a temporary fix until i am able to assign opengl duties to the gpu. note: the function is via the package nvidia-prime)

`
__NV_PRIME_RENDER_OFFLOAD=1 __GLX_VENDOR_LIBRARY_NAME=nvidia python3 /home/username/opt/UVR/UVR.py`

found this at https://wiki.archlinux.org/title/PRIME under subheading 2.2.2

still getting the playsound error though. got the GUI finally though. was able to download a demucs model and now i'm processing something. hopefully this can be helpful to others who get the pyglet error.

**Reply by ste1590** [reply]
Well, you are using Arch, btw. :) When I was a beginner, they taught me not to make a FrankenArch (or FrankenDebian), so when you update or install something, always use the command pacman -Syu. If something breaks, it means that something was already wrong, so you must fix it anyway, haha. Anyway, my main distro is Debian, not Arch but I also like Arch, it's an excellent distro. The main problem with vanilla Arch installation was no glu package, that's why the pyglet error was coming out, at least in my case. Everything else is related to the Python. Just follow my tutorial and it will work correctly. From the moment I installed Arch, it took me 15 minutes to install everything and run UVR with no errors. You have already installed a bunch of packages, I guess it won't hurt but still, you need to install everything that is needed.

EDIT: As I already said, I tested this on VirtualBox. For AMD, there should be no problem, and you can use UVR with CPU processing, not GPU, with no errors. If you have Nvidia it is likely that additional errors are possible but I am not able to recreate it. I have an AMD CPU, and AMD GPU so I can use UVR only with CPU power, it's just fine for my needs but for those of you who have NVIDIA and want to use GPU processing, I imagine there are additional issues that need to be addressed.

**Reply by joshgura** [reply]
UVR is working like a charm now. the file browser leaves something to be desired, but it's forgivable. If I had a life I would have never figured out that my particular problem was because of openGL running on integrated graphics rather than on the GPU.. i eventually gathered that from reading as many posts about pyglet not working, and talking to someone at the pyglet team.
